README.md
Rrobots
=======

rrobots game like (api not strikly compatible)

based on gtk (by Ruiby)


API
===

class template
-----
    class MyThank < Tank
     def tick(time)       # called when state==nill
        ..use thank api..
     end
     def tick_foo(time)   # called when state==foo
        ....use thank api.. 
     end
    end
    MyThank  # !!! Indispensable !!!

api
---
direction are in degree (0..360), positions are in pixel,speed in pixel/tick

    @cdir   cannon direction
    @cradar radar direction
    @a      thank move direction
    x,y     thank position
    @v      thank speed

    turn_radar(delta_direction)  # change radar direction
    turn_cannon(delta_direction) # change cannon direction
    
    pivot_by(delta_direction)    # change moving, direction dont'move cannon & radar direction
    turn(delta_direction)        # change moving direction with cannon and radar direction
    turn_to(direction)           # force current move direction
    mindist?                     # true if game border is reach
    move                         # apply speed to thank position/direction
    
    fire_good?                   # true if there are (other familly) thank in radar direction
    count_bullet                 # number of active bullet (bullet fired by thank and not yet dead)
    fire                         # fire a bullet, done if count_bullet() is less than 6
    
    accelerate(dv)
    tir(condition, state: cible) { action } # move to state name if condition is true
                                          tick_<@state> will be call next tick


thank commented
---------------

    class T < Tank
     def initialize(x,y,coul="#FFAABB") 
        super(x,y,coul) ;  # coul= fill color of thank, x/y : position
        @v= 8              # initial speed
        @cradar=0          # initial directions
        @cible=-1
        turn(rand(100))
      end
      def tick(c)                            # c is incremented on each call==> game time
         pivot_by(rand(45)) if mindist?      # turn if border reached
         turn_radar(0.25)                    # turn radar continuously
         turn(0.5)                           # turn tank
         move()                              
         if  fire_good?                      # radar detect a enemy
           @cdir=(@cradar+@cdir)/2           # turn cannon
           @cible=@cradar                    # memo enemy direction
         end
         fire if (@cible-@cdir).abs<1        # fire a bullet id cannon is in enemy direction
      end
    end
    T


                                          
bugs
----
end of game is not well detected !
@var should not be directly read/modified                                      



CHANGELOG.txt
0.9.6
    rrobots.rb : initiale player
    README.md : initiale player
    README.md : api comments
    README.md : pi comments
    README.md : api documentation
    tanks.rb : creation TbakGenie2
    tank.rb : no traces
    rrobots.rb : slider for choose number of players
    mobile.rb : no trace
    bullet.rb : bullet size
    rrobots.rb : obstacle
    tanks.rb : genie creation
    tank.rb : refactor
    rrobots.rb : end game detection
    mobile.rb : geometry class
    bullet.rb : color
    tanks.rb : anim with time argument
    tank.rb : tick() creation
    rrobots.rb : anim with time argument
    obstacles.rb : anim with time argument
    mobile.rb : tick() creation
    fire.rb : anim with time argument
    bullet.rb : anim with time argument
    tanks.rb : TankEvil creation
    tank.rb : fire definition
    rrobots.rb : acces to source code of a tank
    mobile.rb : angle() : normalyse 
    tanks.rb : Comment for change in bullet.rb : license in header
    tank.rb : LF in bullet.rb
    rrobots.rb : Comment for change in bullet.rb : license in header
    obstacles.rb : LF in bullet.rb
    mobile.rb : Comment for change in bullet.rb : license in header
    fire.rb : LF in bullet.rb
    bullet.rb : license in header
    Rakefile.rb : to rrobots

tank.rb
# Creative Commons BY-SA :  Regis d'Aubarede <regis.aubarede@gmail.com>

# a tank have
#   a move direction, a speed 
#   a connon with a direct
#   a radar with his direction
#   a move of the tank move cannon and radar
#   a turn of tank do same turn to cannonand radar
#
class Tank < Mobile 
  def initialize(x,y,coul="#AA0000")
    super
    @state=""
    @count_bullet=0
  end
  # draw radar, body+wheels, cannon
  def draw(w,cr)
    w.draw_line([@x,@y,@x+1000*cos(@cradar),@y+1000*sin(@cradar)],"#007000") if @cradar!=0
    w.rotation(@x,@y,deg2rd(@a)) { 
      w.draw_rectangle(-5,-5,10,10,0,@coul,@coul) 
      w.draw_line([-7,-5,+7,-5],"#000000",2) 
      w.draw_line([-7,+5,+7,+5],"#000000",2) 
    }
    w.draw_line([@x,@y,@x+10*cos(@cdir),@y+10*sin(@cdir)],"#000000")
  end
  def turn_cannon(dr) @cdir=alimit(@cdir+dr)       end
  def turn_radar(dr)  @cradar=alimit(@cradar+dr)   end
  def dead_bullet()   @count_bullet-=1             end
  def count_bullet()  @count_bullet                end
  def fire()          (@count_bullet+=1 ; $app.creatBullet(self,@x,@y,@cdir)) if @count_bullet<6 end
  def fire_good?
    $app.each_tank(self).select { |t| inline?(t,10) }.size>0
  end
  def tir(condition,option,&blk)
     if condition  
        #puts "tir to #{option[:state]}..."
        to_state(option[:state])
        blk.call if blk
     end
  end
end


tanks.rb
# Creative Commons BY-SA :  Regis d'Aubarede <regis.aubarede@gmail.com>

class TankGamer < Tank ; end

class TankGenie2 < TankGamer
 def initialize(x,y,coul="#FFAA66") 
    super(x,y,coul) ; 
    @v,@cradar,@cible,@tc= 11,0,-400,9999999
    @s=rand(1)+1
    turn(rand(360))
  end
  def tank?() true end
  def common
     pivot_by(45) if mindist?
     turn_radar(@s*2)
     @cdir=@cradar-@s*2
     if fire_good? 
		@cible=@cradar
     end
     move()
  end
  def tick(c)
     common
     tir((@cible-@cdir).abs<2 && count_bullet()<4 ,state: :a) { 	
		 fire 
	 }
     tir((@cible-@cdir).abs<4 && count_bullet()>=4 ,state: nil) { 	
       pivot_by -10
       accelerate  1.3
	 }
     @v = 11 if @v>15
  end
  def tick_a(c)
    common
    @a=( (8.0*@a+@cible)/9.0 )
    if c%2>=0
      fire
    end
    tir( (@cible-@cdir).abs>3,state: :b) do
       @s *= -1
       @cible=999999999
       @v=11
    end
  end
  def tick_b(c)
    common
    accelerate 1.1
    fire if c%2==0
    tir( c>6,state: nil) do
        @s *= -1
        @v=11
    end
  end
end

class TankGenie < TankGamer
 def initialize(x,y,coul="#9999FF") 
    super(x,y,coul) ; 
    @v,@cradar,@cible,@tc= 11,0,-400,9999999
    @s=1
    turn(rand(360))
  end
  def tank?() true end
  def common
     pivot_by(45) if mindist?
     turn_radar(@s*2)
     @cdir=@cradar-@s*0.1*(-1)
     turn_cannon(1)
     if fire_good?
		@cible=@cradar
     end
     move()
  end
  def tick(c)
     common
     tir((@cible-@cdir).abs<2,state: :a) { @v*=-1; fire }
  end
  def tick_a(c)
    common
    fire if c%3==0
    tir( (@cible-@cdir).abs>3,state: :b) do
       @cible=999999999
       @v=11
    end
  end
  def tick_b(c)
    common
    fire if c%3==0
    tir( c>6,state: nil) do
        @s *= -1
        @v=11
    end
  end
end

class TankEvil < TankGamer
 def initialize(x,y,coul="#FFAABB") 
    super(x,y,coul) ; 
    @v= 8
    @cradar=0
    @cible=-1
    turn(rand(100))
  end
  def tank?() true end
  def anim(c)
     pivot_by(rand(45)) if mindist?
     turn_radar(0.25)
     turn(0.5)
     move()
     if  fire_good?
       @cdir=(@cradar+@cdir)/2
       @cible=@cradar
     end
     fire if (@cible-@cdir).abs<1
  end
end

class TankManuel < TankGamer

  def initialize(x,y,coul="#FF0000") super(x,y,coul) ; end
  def anim(c)
    turn_radar(1)
    accelerate(0.95)
    move()
  end
  def tank?() true end
end

class TankSimple < TankGamer
  def initialize(x,y,coul="#FF4444") 
    super(x,y,coul) ; 
    @a=rand(0..360)
    @v= 3
    @cradar=@a
  end
  def tank?() true end
  def anim(c)
     pivot_by(rand(15)) if mindist?
     turn(1)
     move()
     (fire;accelerate(1.1);turn(-1)) if fire_good?
     accelerate(0.999)
  end
end


class TankDuck3 < TankGamer
  def initialize(x,y,coul="#44FF44") 
    super(x,y,coul) ; 
    @v= 3
    @cradar=0
  end
  def tank?() true end
  def anim(c)
     turn_to(4)
     move()
  end
end

mobile.rb
# Creative Commons BY-SA :  Regis d'Aubarede <regis.aubarede@gmail.com>
class Geom
  PI=Math::PI
  def alimit(a)
      a -= 360 while a > 360
      a += 360 while a < 0
      a
  end
  def cos(a) Math.cos((PI*a)/180.0) end
  def sin(a) Math.sin((PI*a)/180.0) end
  def atan2(dy0,dx0) rd2deg(Math.atan2(dy0,dx0)) end
  def deg2rd(a) (PI*a)/180.0 end
  def rd2deg(a) (180.0*a)/PI end
  def crop(min,a,max)
     return min if a<min
     return max if a>max
     a
  end
  def inline?(t,epsilon)
      return(false) if t.class == self.class
      dx,dy=t.x-x, t.y-y
      h=Math.sqrt(dx*dx+dy*dy)
      b=atan2(dy,dx)
      aa=alimit(b-@cradar)
      dist= (h*sin(aa)).abs
      is=(dist<epsilon)&& aa<90
      #p({:ee => self.class, :atan2 => b, :adiff => aa}) if is && self.kind_of?(T)
      is
  end
end

class Mobile < Geom
  attr_accessor :x,:y,:cdir,:cradar,:a
  def tank?() false end
  def obstacle?() false end  
  def initialize(x,y,coul="#AA0000")
    @x,@y,@dx,@dy,@coul=x,y,0,0,coul
    @count,@v,@a,@cdir,@cradar=0,0,0,0,0
    @state=""
  end
  def draw(w,cr)
      w.draw_rectangle(x-5,y-5,10,10,0,@coul,@coul) 
  end
  
  ######################## Movement ##################
  
  def intersect(other)
    ( (x-other.x)*(x-other.x)+ (y-other.y)*(y-other.y) ) < 8*8
  end
  def accelerate(c) @v= @v.abs<1 ? 1 : @v*=c end
  def pivot_by(c)   @a=alimit(@a+c)          end  
  def turn(c) 
     @a=alimit(@a+c)
     @cdir=alimit(@cdir+c)
     @cradar=alimit(@cradar+c)
  end
  def turn_to(a)
     da=alimit(a-@a)
     @cdir+= da
     @cradar+= da
     @a=a
  end
  def mindist?()
     x<5 || y<5 || x>($W-5) || y>($H-5)
  end
  def to_state(s)
    @state=s ? "_"+s.to_s : ""
    Message.alert("State unknown #{s} in class #{self.class}") unless self.respond_to?("tick#{@state}")
  end
  def dead() end
  def move()
    nx= @x+@v*cos(@a)/10
    ny=@y+@v*sin(@a)/10
    @x= crop(0,nx,$W)
    @y= crop(0, ny,$H)
    if $app.obstacle?(self)
      @v=-@v
    end
    @v=0 if nx!=@x || ny!=@y
  end
  
  ######################### Private  
  
  def childr?(t)      t==@parent end
  def pxy() 
    puts " #{@coul}: @ %4f/%4f v=%4f}/%4f  dir=#{@a} canon=#{@cdir} radar=#{@cradar} #{self.class.to_s}" % [@x,@y,@dx,@dy]
  end
  
  
  def _anim()
    if self.respond_to?(:tick)
      s=@state
      send("tick#{@state}",@count)
      if @state!=s
        @count=0
      end
    else
      anim(@count)
    end
    @count+=1
  end
end


Rakefile.rb
# Creative Commons BY-SA :  Regis d'Aubarede <regis.aubarede@gmail.com>

#################################################################
#  Rakefile : git commit/add ; git push, test/spec & gem push
#################################################################
#
# Usage:
# > rake commit  # commit in local directory and push to repository
# > rake gem     # make gem,and test it, and push it to gemcutter
# > rake travis  # travis CI run commands
#
#################################################################
#  Requirements :
#   version : X.Y.Z  
#           X : to be hand-modified 
#           Y : incremented at each 'rake gem'
#           Z : increment on each file 'rake commit'
#   auto commit all changed file, sking comment for each
#   log all commment in CHANGELOG?Txt file
#   increment VERSION content with version numbering rule
#     so CHANGELOG.txt is updated with each comment/version change
#     so VERSION is updated in each commit (Z part) and each gem build (Y part) 
#   
# commit:
#   Use 'git status' output for each new/modified file,
#   with asking some comment
#   (no add/commit of comment is empty)
#
# gem:
#   execute rspec if specs exists,
#   make a gem, install it, execute samples/test.rb in it, if exist
#   ask to operator if execution is ok and : 
#      increment version
#      update change log/version
#      do a 'gem push' and a 'gem yanked'
#    make a coffe
#
#################################################################

FIGNORES=%w{VERSION CHANGELOG.txt .gitignore}
NAME= Dir.pwd.gsub(File.dirname(Dir.pwd)+'/',"")

Rake.application.options.trace = false

def push_changelog(line)
  b=File.read('CHANGELOG.txt').split(/\r?\n/) 
  b.unshift(line)
  File.open('CHANGELOG.txt','w') {|f| f.write(b[0..500].join("\n")) }
end
def changelog_push_currrent_versions()
  version=File.read('VERSION').strip
  b=File.read('CHANGELOG.txt').split(/\r?\n/) 
  b.unshift(version)  
  File.open('CHANGELOG.txt','w') {|f| f.write(b.join("\n")) }
  yield rescue p $!.to_s
  b.shift  
  File.open('CHANGELOG.txt','w') {|f| f.write(b.join("\n")) }
end

def change_version()
  a=File.read('VERSION').strip.split('.')[0..2]
  yield(a)
  version=a.join('.') 
  File.open('VERSION','w') {|f| f.write(version) }
  version
end
def verification_file(fn)
  return unless File.extname(fn)==".rb"
  content=File.read(fn)[0..600]
  unless content =~ /BY-SA/ 
    puts "\nFile #{fn} seem not contain licenses data (CC BY-SA)\n"   
    exit(0)
  end
end
#############################################################
#               Comment each file add/modified             ##
#############################################################


desc "commit file changed and created"
task :commit_status do
  `git status -s`.split(/\r?\n/).each do |line|
  words=line.split(/\s+/)
  case line
    when /^ M /
      filename=words[2]
      next if FIGNORES.include?(filename)
      system("git","diff",filename)
      print("Comment for change in #{filename} : ")
      comment=$stdin.gets
      if comment && comment.chomp.size>0
          comment.chomp!
          (puts "Abort!";exit!)   if comment=~/^a(b(o(r(t)?)?)?)?$/
          verification_file(filename)
          sh "git commit #{filename} -m \"#{comment.strip}\"" rescue 1
          push_changelog("    #{File.basename(filename)} : #{comment}")
          $changed=true
      end
    when /^\?\?/
      filename=words[1]
      print("Comment for new file in #{filename} : ")
      comment=$stdin.gets.chomp
        (puts "Abort!";exit!)   if comment=~/^a(b(o(r(t)?)?)?)?$/
      if comment =~ /^y|o/i
        verification_file(filename)
        sh "git add #{filename}"
        sh "git commit #{filename} -m \"creation\"" rescue 1
        $changed=true
      end
  end
  end
end


#############################################################
#  before commit
#############################################################

desc "job before commit"
task :pre_commit do
  sh "cls" if RUBY_PLATFORM =~ /(win32)|(mingw)/i 
  puts <<EEND2


--------------------------------------------------------------------
                 Commmit and push #{NAME}
--------------------------------------------------------------------
EEND2
  
  #sh "giti"
  $changed=false
end

#############################################################
#  after commit
#############################################################
desc "job after local commit done: push to git repo"
task :post_commit do
  if $changed
    $version=change_version { |a| a[-1]=(a.last.to_i+1) }  
    sh "git commit VERSION -m update"
    changelog_push_currrent_versions {
      sh "git commit CHANGELOG.txt -m update"
      sh "git push"
      puts "\n\nNew version is #{$version}\n"
    } 
  else
    puts "no change!"
  end
end

#############################################################
#   commit
#############################################################
desc "commit local and then distant repo"
task :commit => [:pre_commit,"commit_status",:post_commit]


#############################################################
#  gem build & push
#############################################################
desc "make a gem and push it to gemcutter"
task :gem => :commit do
  puts <<EEND


--------------------------------------------------------------------
       make gem, test it localy, and push gem #{NAME} to gemcutter
--------------------------------------------------------------------
EEND
  ruby "samples/make_doc.rb","1"
  sh "git commit doc.html -m update"
  sh "git push"
  $version=change_version { |a| 
      a[-2]=(a[-2].to_i+1) 
      a[-1]=0 
  }   
  puts "New version ==>> #{$version}"
  l=FileList['*.gem']
  l.each { |fn| rm fn }
  gem_name="#{NAME}-#{$version}.gem"
  push_changelog  "#{$version} : #{Time.now}"
  sh "gem build #{NAME}.gemspec"
  Rake::Task["test"].execute    if File.exists?("samples/test.rb")
  sh "gem push #{gem_name}"
  l.each { |fn|
      ov=fn.split('-')[1].gsub('.gem',"")
    sh "gem yank -v #{ov} #{NAME}"
  }
end
#############################################################
#  execute tests 
#############################################################

desc "test the current version of the framework by installing the gem and run a test programme"
task :test do
 if File.exists?("spec/test_all.rb")
  system("rspec spec/test_all.rb")
 end
 if File.exists?("samples/test.rb")
   cd ".."
   mkdir "#{NAME}Test" unless File.exists?("#{NAME}Test")
   nname="#{NAME}Test/test.rb"
   content=File.read("#{NAME}/samples/test.rb").gsub(/require_relative/," require").gsub('../lib/','')
   File.open(nname,"w") { |f| f.write(content) }
   sh "gem install #{FileList["#{NAME}/#{NAME}*.gem"][-1]}"
   ruby nname rescue nil
   cd NAME
   print "\n\nOk for diffusion ? "
   rep=$stdin.gets
   raise("aborted!") unless rep && rep =~ /^y|o|d/
 end
end

#############################################################
#  travis command 
#############################################################

task :travis do
 puts "Starting travis test..."
 cmd="bundle exec rspec spec/test_all.rb"
 system("export DISPLAY=:99.0 && #{cmd}")
 #system(cmd)
 raise "rspec failed!" unless $?.exitstatus == 0
end




rrobots.rb
# Creative Commons BY-SA :  Regis d'Aubarede <regis.aubarede@gmail.com>

#require_relative '../Ruiby/lib/Ruiby'
require 'Ruiby'
require_relative 'mobile'
require_relative 'tank'
require_relative 'bullet'
require_relative 'fire'
require_relative 'tanks'
require_relative 'obstacles'

$W=500
$H=500

def source_class(klass)
  lsm = (klass.methods(false).map{|m| klass.method(m)} +
             klass.instance_methods(false).map{|m| klass.instance_method(m)}).
            map(&:source_location).
            compact
  file,start,count=lsm.group_by{|sl| sl[0]}.map do |file, sls|
    lines = sls.map{|sl| sl[1]}.sort
    count = lines.last-lines.first
    line = lines.min
    {file: file, count: count, line: line}
  end.sort_by!{|fc| -fc[:count]}.map{|fc| [fc[:file], fc[:line], fc[:count]]}[0]
  
  lines=File.read(file).split(/\r?\n/)
  nol_start=start.downto(0) { |nol| break(nol) if lines[nol] =~/^class\b/ && lines[nol] !~ /<</}
  nol_end=(start+count).upto(lines.size) { |nol| break(nol) if lines[nol]=~/^end/ }
  if nol_start && nol_end && nol_end>nol_start && nol_end<= lines.size
    lines[nol_start..nol_end].join("\n")
  else 
    raise("no class found")
  end
end

class Test
  class << self
     def test(string)
       eval string
     rescue Exception => e
       Message.alert(e.to_s)
     end
  end
end

Ruiby.app width: 300+$W, height: $H+50, title: "RRobots" do
  src=Ruiby.stock_get("source",<<'EEND')
class T < Tank
  def initialize(x,y,coul="#4444FF") 
    super(x,y,coul) ; 
    @v= 10
    @a=@cdir=@cradar=0
  end
  def tank?() true end
  def tick(c)
     turn_to(2)
     pivot_by(45) if mindist?
     fire if fire_good? 
     move()
  end
end
T
EEND
  $app=self
  @announce=""
  @status_game=0
  @ltank =[]
  def_style "* { font-size: 10px;}"
  def_style "button { background-color: #FF0000;}"
  stack do
  flow do
    stack do
      @ed=source_editor(width:400, height:500) 
      @ed.editor.buffer.text=src
      flowi do
         button("Syntax test ")  { 
           Ruiby.stock_put("source",@ed.editor.buffer.text)
           Test.test @ed.editor.buffer.text 
         }
         button("Replace current player") { 
           Ruiby.stock_put("source",@ed.editor.buffer.text)
           begin
              eval @ed.editor.buffer.text 
           rescue Exception => e
              error(e)
           end
        }
      end
    end if true
    stacki do
      @cv=sloti(canvas($W,$H) do 
        on_canvas_draw { |w,cr|   
          w.init_ctx("#000000","#005050",1)
          @ltank.each { |t| t.draw(w,cr) } 
          w.draw_text(10,40,@announce,3,"#FF5577") if @announce.size>0
        }
      end)
      stack {
        ctx=  make_StockDynObject("ctx",{"nb_player" => 1, "nb_robots" =>  4})
        lTanks=ObjectSpace.each_object(::Class).select {|klass| klass < TankGamer ? klass: nil}.map {|c|c.to_s}
        flowi { labeli("Play with :",width:150) ; @cb=sloti(combo(lTanks,0))  }
        flowi { labeli("nb Players:",width:150) ; sloti(islider(ctx.nb_player,min:1,max:30,by:2)) }
        flowi { labeli("nb Robots:",width:150)  ; sloti(islider(ctx.nb_robots,min:1,max:30,by:2) ) }
        flowi {
          regular
          button("view") { edit(source_class(eval @cb.get_selection()[0])) }
          button("Go") { run_game(@cb.get_selection()[0],ctx.nb_player.value,ctx.nb_robots.value) }
        }
      }
    end
  end
    buttoni("reload") { 
      load __FILE__ 
      load "mobile.rb"
      load "tank.rb"
    }
  end
  anim(40) {  
    @ltank.each { |t| t._anim } if @status_game <2
    @announce=case @status_game
        when 0 then "Waiting to Start"
        when 1 then ""
        when 2 then "Bravo !!!"
        when 3 then "Game over ;("
    end
    @cv.redraw 
    end_game() if  @ltank.size==1
  } 
  def run_game(klass_namame,nbp,nbr) 
    klass= (eval(klass_namame) rescue nil)
    return unless  klass
    Ruiby.stock_put("source",@ed.editor.buffer.text)
    begin
      tclass = eval @ed.editor.buffer.text 
    rescue Exception => e
       error e
       return
    end
    @ltank =[]
    nbp.to_i.times { @ltank << @player=tclass.new(rand(20..$W-20),rand(20..$H-20)) }
    nbr.to_i.times { @ltank << klass.new(50+rand($W-50),50+rand($H-50)) }
    #4.times { @ltank << Obstacle.new(rand(0..$W),rand(0..$H),rand(3..15)) }
    @status_game=1
  end
  class << $app
    def creatBullet(o,x,y,dir)  @ltank << Bullet.new(o,x,y,dir)  end
    def kill(t) 
      unless t.obstacle? 
        fire(t.x,t.y)
        dead(t)
      end
    end
    def dead(o)       
      o.dead
      @ltank.delete(o) 
    end
    def fire(x,y)     @ltank << Fire.new(x,y) end
    def obstacle?(o) 
      @ltank.any? { |t|  t!=o && t.obstacle? && t.intersect(o) }  
    end
    def test_collision(o,parent)
      dead(o) if @ltank.any? { |t|  (t!=o && ! o.childr?(t) && (t.tank? || t.obstacle?) && t.intersect(o)) ? (kill(t); true) : false }  
    end
    def each_tank(o)
       if block_given?
         @ltank.each { |t|  if t!=o && t.tank? then  yield end }  
       else
         @ltank.select { |t|  (t!=o && t.tank? ) }.each  
       end
    end
    def end_game()
     ok= true || (@ltank.size==1 && @ltank.first==@player)
     @status_game = ok ? 2 : 3
     #@ltank=[]
    end
  end
end
